homeassistant/components/apcupsd/sensor.py

```diff
@@ -334,16 +335,20 @@
     "stesti": SensorEntityDescription(
         key="stesti",
         translation_key="self_test_interval",
+        device_class=SensorDeviceClass.DURATION,
     ),
     "timeleft": SensorEntityDescription(
         key="timeleft",
         translation_key="time_left",
         state_class=SensorStateClass.MEASUREMENT,
+        device_class=SensorDeviceClass.DURATION,
     ),
```

Comment:
Missing unit of measurement

Reply:
These two sensors would not have a consistent unit of measurement since the underlying daemon does time conversion internally. Sometimes it reports "7 days" and sometime it reports "30 minutes", which is why we have `infer_unit` helper function within the integration that sets the field dynamically.

We should probably set a fixed unit of measurement here and internally convert the units to the fixed unit of measurement. Does HA expose such helper functions? (it definitely does have unit conversions, but I'm not sure if it also exposes those functions for integrations to use). Anyway, that should be handled in a separate PR.

Reply:
But in order to add a device class, the entity should have a UoM. 
The UoM of a sensor can change at runtime, as long as it's one of the UoMs of within the device class.

Reply:
I see. Updated with a default UoM.

Reply:
But does this now work correctly if it shows 3 second or 5 days like you mentioned?

Reply:
> The UoM of a sensor can change at runtime

Would HA handle this gracefully? Let's say if the daemon reports 5 minutes, and our integration sets the UoM to be minutes and the value to be 5. At next pull the daemon reports 3 seconds, if our integration updates the UoM (e.g., via setting `_attr_native_unit_of_measurement`), would HA automatically convert the history values (5 minutes -> 300 seconds) or the old value would now be incorrectly recorded as "5 seconds"? This was also the previous motivation to have a canonical native UoM for the sensors.

But that actually made me realize that the old logic here (https://github.com/home-assistant/core/blob/dev/homeassistant/components/apcupsd/sensor.py#L468-L470) that we inherited from the sync version of the integration may be flawed: it doesn't handle UoM changes at all. 

We should probably fix that before we refine the descriptions here.

Reply:
Investigated a littler deeper here into the APCUPSD source code:

For "STESTI": it could output arbitrary strings with varying units. Although the test code always shows that this is a pure value where the unit is "HOURS". But we've had user reports in HA that it can also output something like "7 days" (perhaps in certain models). 
ref: https://fossies.org/dox/apcupsd-3.14.14/apcstatus_8c_source.html#l00327 

For "CUMONBATT": the unit is always seconds.
ref: https://fossies.org/dox/apcupsd-3.14.14/apcstatus_8c_source.html#l00280

For "TIMELEFT": the unit is always minutes.
ref: https://fossies.org/dox/apcupsd-3.14.14/apcstatus_8c_source.html#l00149

For "TONBATT": the unit is always seconds.
ref: https://fossies.org/dox/apcupsd-3.14.14/apcstatus_8c_source.html#l00279

So, this shows that TIMELEFT and CUMONBATT have fixed UoM, but STESTI is a weird one. I have removed the UoM and device class for STESTI and leave that for future PRs (for figuring out the conversions), and kept the fixed UoM for CUMONBATT and TIMELEFT. 

Let me know your thoughts (and also if you know more information of the graceful handling of units in HA) 😃

Reply:
It does handle it nicely for the long term statistics, but I'm not sure if it does for automations and such. Let me try and consult some others. Feel free to remind me.

Reply:
Gentle reminder :) it seems that the only weird one is STESTI and others have fixed UoM (confirmed from the source code of apcupsd), and this PR is now purely refining the units of the fixed UoMs.

Further refinement for STESTI will be made in follow-up PRs.